From a913c2a1c3ed9a07ee778a6275f2a8b4e3e230da Mon Sep 17 00:00:00 2001
From: "Lin.TommySC" <lin.tommysc@inventec.com>
Date: Tue, 8 Jun 2021 09:30:32 +0000
Subject: [PATCH] Force zero padding when setting password

Symptom/Reason :
    - Create password in 16 byte password size may fail sometimes

Root Cause:
    - Passwords are not zero padded occasionally

Solution/Change:
    [phospher-ipmi-host]
        - Force 0 padding at end of password

Entry Test:
    - Create new user password
        ipmitool -I lanplus -C 17 -p 2623 -U root -P 0penBmc -H 172.17.0.3 user set name 13 KzcoXNTg
    - Set user password
        ipmitool -I lanplus -C 17 -p 2623 -U root -P 0penBmc -H 172.17.0.3 user set password 13 0penBmc10penBmc2Bmc3 16
    - Grant user access
        ipmitool -I lanplus -C 17 -p 2623 -U root -P 0penBmc -H 172.17.0.3 channel setaccess 1 13 ipmi=on privilege=4
    - Enable user account
        ipmitool -I lanplus -C 17 -p 2623 -U root -P 0penBmc -H 172.17.0.3 user enable 13
    - Use new password to verify
        ipmitool -I lanplus -C 17 -p 2623 -U KzcoXNTg -P 0penBmc10penBmc2 -H 172.17.0.3 sel info
---
 user_channel/usercommands.cpp | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/user_channel/usercommands.cpp b/user_channel/usercommands.cpp
index bda4027..40d3001 100644
--- a/user_channel/usercommands.cpp
+++ b/user_channel/usercommands.cpp
@@ -392,8 +392,12 @@ Cc ipmiSetUserPassword(ipmi_netfn_t netfn, ipmi_cmd_t cmd,
     }
     if (req->operation == setPassword)
     {
+        // Tommy. Add zero padding to end of password.
+        uint8_t password[maxIpmi20PasswordSize] = {0};
+        std::memcpy(password, req->userPassword, passwordLength);
+        // Tommy (End)
         return ipmiUserSetUserPassword(
-            req->userId, reinterpret_cast<const char*>(req->userPassword));
+            req->userId, reinterpret_cast<const char*>(password));
     }
     else if (req->operation == enableUser || req->operation == disableUser)
     {
-- 
2.7.4

